source/src/AnalysisHelper.cc

```diff
@@ -0,0 +1,179 @@
+/// \file AnalysisHelper.cc
+/*
+ *
+ * AnalysisHelper.cc source template automatically generated by a class generator
+ * Creation date : mar. juil. 1 2014
+ *
+ * This file is part of DQM4HEP libraries.
+ *
+ * DQM4HEP is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * based upon these libraries are permitted. Any copy of these libraries
+ * must include this copyright notice.
+ *
+ * DQM4HEP is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with DQM4HEP.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @author Tom Coates
+ * @copyright CNRS , IPNL
+ */
+
+// -- dqm4hep headers
+#include <dqm4hep/AnalysisHelper.h>
+#include <typeinfo>
+
+namespace dqm4hep {
+
+  namespace core {
+
+    float AnalysisHelper::mean(MonitorElementPtr pMonitorElement, float percentage = 1.0)
+    {
+
+      std::string ObjectType;
+
+      float result = 0.0;
+
+      TObject *ThisObject = pMonitorElement->objectTo<TObject>();
+
+      if (pMonitorElement->type() == "TH1F" || pMonitorElement->type() == "TH1I") {  // How many types of TH1 histogram do we have to recognise? TH1F, TH1I, ...?
```

Comment:
Here you can use something like : 
```cpp
if(nullptr != pMonitorElement->objectTo<TH1>() and nullptr == pMonitorElement->objectTo<TH2>) {
  // ...
}
```
Same for the other conditions below

Reply:
Casting either a TGraph to a TH1, or a TH1 to a TGraph, both produce non-null pointers. This is what I have been using:

      if (nullptr != pMonitorElement->objectTo<TGraph>()) {
	ObjectType = "TGraph";
      }
      else if (nullptr != pMonitorElement->objectTo<TH1>()){
	ObjectType = "TH1";
      }
      else {
	dqm_error("The monitor element {0} of type {1} could not be recognised.", pMonitorElement->name(), pMonitorElement->type());
	throw StatusCodeException(STATUS_CODE_FAILURE);
      }

But this just results in the object being classified as whatever the first condition is, which then produces segfaults when I use a TH1-only method on a TGraph object and vice versa. In the meantime I've been adding a comparison of `pMonitorElement->type()` against a string, but is there a more foolproof way of determining the object type using casting? Can a TGraph be cast to a TH2?

Reply:
For this purpose we're only interested in whether the cast itself returns a nullptr though, and it seems in any of these cases they do return a valid pointer, even if using it causes a segfault?

Reply:
I think I observed this problem at some point when implementing other QTest. I believe this was due to a wrong cast in the `MonitorElement::objectTo<T>` method. This is normally fixed in the head version. Have you already rebased your code on master ?

Reply:
My branch isn't rebased to the master yet, I'll do that and see if the issue persists.